### PR TITLE
Add clang-offload-deps tool for generating dependence files for offload targets

### PR DESCRIPTION
**Pull request by sndmitriev**
This tool is intended to be used by the clang driver for offload linking with
static offload libraries. It takes linked host image as input and produces
bitcode files (one per offload target) containing references to symbols that
must be defined in the target images. Each dependence bitcode file then is
expected to be compiled to an object by the driver using appropriate target
toolchain, and dependence object added to the target linker as input together
with the other inputs. References to the symbols in dependence object should
ensure that target linker pulls in necessary symbol definitions from the input
static libraries.

Signed-off-by: Sergey Dmitriev <serguei.n.dmitriev@intel.com>

### CODE REVIEW

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (kbobrovs) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:54]: Usually multiple outputs within the driver are handled either via file lists (TY_Tempfilelist) or file tables (TY_Tempfiletable).  Driver implementation would easier and more inline with other multi-output tools if this tool could generate a file list listing all the outputs separated by \n.
@mdtoguchi - what do you think?

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (kbobrovs) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:112]: Looks like `!Kind.empty() &&` is not needed

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (kbobrovs) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:216]: I think from code maintenance perspective it is better to have single mechanism for SPIRV and non-SPIRV. Will the second work for SPIRV target as well? If yes, can the first one be discarded?

**Comment by bader** [comment]
@AlexeySachkov, FYI.

**Comment by bader** [comment]
@sndmitriev, please, add yourself as a code owner for this tool to https://github.com/intel/llvm/blob/sycl/.github/CODEOWNERS.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (sndmitriev) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:216]: Well, LLVM IR just allows to create a reference without adding any additional symbols, I do not see why we should not use this feature. I would rather discard the second part if we had a way to add a reference without adding extra symbols, but looks like this is the only way to do it.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (bader) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:146]: This seems to implement `std::find_if` logic, so I would use standard library function instead of for-loop.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (bader) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:157]: Is it expected that we consider files w/o `.tgtsym` as bad input? Will it be a problem to produce an empty list of dependencies for such files?

File: clang/test/Driver/clang-offload-deps.c
Comment (sndmitriev) [reply clang/test/Driver/clang-offload-deps.c:12]: It won't work because this test does not check the full output. Here is how it looks like

OVERVIEW: A tool for creating dependence bitcode files for offload targets. Takes
host image as input and produces bitcode files, one per offload target, with
references to symbols that must be defined in target images.

USAGE: clang-offload-deps [options] <input file>

OPTIONS:

Generic Options:

  --help             - Display available options (--help-hidden for more)
  --help-list        - Display list of available options (--help-list-hidden for more)
  --version          - Display the version of this program

clang-offload-deps options:

  --outputs=<string> - [<output file>,...]
  --targets=<string> - [<offload kind>-<target triple>,...]

Comment (bader) [reply clang/test/Driver/clang-offload-deps.c:12]: I see. Probably we should check the test in OVERVIEW section with `CHECK-HELP-NEXT` and keep other checks as is, if we don't want to check the full output.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (sndmitriev) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:157]: Such inputs are not expected to cause any errors. I will update LIT test to check that we produce an empty list for such inputs.

File: clang/test/Driver/clang-offload-deps.c
Comment (sndmitriev) [reply clang/test/Driver/clang-offload-deps.c:12]: Okay, I will add -NEXT to OVERVIEW section.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (mdtoguchi) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:16]:
```suggestion
/// object by the driver using the appropriate offload target toolchain. This
/// dependence object is added to the target linker as input together with the
```

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (sndmitriev) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:146]: It is indeed close to `find_if`, but using `find_if` here would be problematic because predicate would be required to handle `Error` which can be returned by `Section.getName()` call (see line 148).

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (mdtoguchi) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:54]: I would expect that the outputs match the targets, which is in-line with the bundler.  I'm OK with how things are.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (mdtoguchi) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:205]:
```suggestion
      // SPIRV linking is done on LLVM IR inputs, so we can use special
```

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (bader) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:146]: Okay. Let's ignore this suggestion.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (bader) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:157]: Right... misread the code. I see that error is reported for empty `.tgtsym` sections.

**Comment by sndmitriev** [comment]
Done

File: sycl/CMakeLists.txt
Comment (sndmitriev) [reply sycl/CMakeLists.txt:283]: Yes, this tool is not integrated into the compilation flow yet.

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp
Comment (kbobrovs) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:215]: AFAIU, both achieve the same result - letting linker know the deps - and only the second can handle all targets. Hence my suggestion. This is a nit anyway.

**Review by bader — DISMISSED** [review]
Thanks!

**Review by kbobrovs — APPROVED** [review]

**Review by romanovvlad — DISMISSED** [review]

**Comment by sndmitriev** [comment]
Does this patch need any additional changes to be merged?

**Comment by romanovvlad** [comment]
@sndmitriev  There is merge conflict in .github/CODEOWNERS, could you please resolve it?

**Comment by sndmitriev** [comment]
> @sndmitriev There is merge conflict in .github/CODEOWNERS, could you please resolve it?

Done.

**Review by bader — APPROVED** [review]

**Comment by sndmitriev** [comment]
I have resolved conflict. Does this change need anything else to be merged?

**Comment by romanovvlad** [comment]
> 
> 
> I have resolved conflict. Does this change need anything else to be merged?

Github UI says:
Waiting on code owner review from intel/llvm-reviewers-runtime, AGindinson, mdtoguchi, elizabethandrews, Fznamznon, and/or premanandrao.

**Review by romanovvlad — APPROVED** [review]

**Review by mdtoguchi — APPROVED** [review]
LGTM

**Review by elizabethandrews — COMMENTED** [review]
I took a look at this patch. The patch looks ok to me but I am unfamiliar with this and so I don’t think I’m the right person to review this. @kbobrovs @bader could you please take a look?

**Pull request merged**
